pepper-apis/src/main/java/org/broadinstitute/ddp/route/CreateUserActivityUploadRoute.java

```diff
@@ -0,0 +1,125 @@
+            FormActivityDef activityDef = activityStore.findActivityDef(handle, studyGuid, activityDto, versionDto)
+                    .orElseThrow(() -> new DDPException("Could not find activity definition for instance " + instanceGuid));
+
+            boolean isInstanceReadOnly = ActivityInstanceUtil.isReadonly(
+                    activityDef.getEditTimeoutSec(), instanceDto.getCreatedAtMillis(),
+                    instanceDto.getStatusType().name(), activityDef.isWriteOnce(), instanceDto.getReadonly());
+            if (!isStudyAdmin && isInstanceReadOnly) {
+                String msg = "Activity instance " + instanceGuid + " is read-only, no file upload will be authorized";
+                LOG.warn(msg);
+                throw ResponseUtil.haltError(response, 422, new ApiError(ErrorCodes.ACTIVITY_INSTANCE_IS_READONLY, msg));
+            }
+
+            String questionStableId = payload.getQuestionStableId();
+            QuestionDef questionDef = activityDef.getQuestionByStableId(questionStableId);
+            if (questionDef == null) {
+                String msg = "Could not find question with stable id " + questionStableId;
+                LOG.warn(msg);
+                throw ResponseUtil.haltError(response, 404, new ApiError(ErrorCodes.QUESTION_NOT_FOUND, msg));
+            } else if (questionDef.getQuestionType() != QuestionType.FILE) {
+                String msg = "Question " + questionStableId + " does not support file uploads";
+                LOG.warn(msg);
+                throw ResponseUtil.haltError(response, 422, new ApiError(ErrorCodes.NOT_SUPPORTED, msg));
+            }
+
+            boolean isQuestionReadOnly = QuestionUtil.isReadonly(handle, questionDef, instanceDto);
+            if (!isStudyAdmin && isQuestionReadOnly) {
+                String msg = "Question " + questionStableId + " is read-only, no file upload will be authorized";
+                LOG.warn(msg);
+                throw ResponseUtil.haltError(response, 422, new ApiError(ErrorCodes.QUESTION_IS_READONLY, msg));
+            }
+
+            User operatorUser = handle.attach(UserDao.class).findUserByGuid(operatorGuid)
+                    .orElseThrow(() -> new DDPException("Could not find operator with guid " + operatorGuid));
+
+            String prefix = String.format("%s/%s/%s", studyGuid, userGuid, instanceDto.getActivityCode());
+            return service.authorizeUpload(
+                    handle, instanceDto.getStudyId(), operatorUser.getId(), instanceDto.getParticipantId(), prefix,
+                    payload.getMimeType(), payload.getFileName(), payload.getFileSize(), payload.isResumable());
+        });
+
+        if (result.isExceededSize()) {
+            String msg = "File size exceeded maximum of " + service.getMaxFileSizeBytes() + " bytes";
+            LOG.warn(msg);
+            throw ResponseUtil.haltError(response, 422, new ApiError(ErrorCodes.NOT_SUPPORTED, msg));
```

Comment:
Consider using HttpStatus.SC_UNPROCESSABLE_ENTITY , etc. where we have integers for status codes.

Reply:
Done! d471b88524f8947ac322622378d83aa8fd0e9b5c